File: Algorithms/Mathematics/Java/karatsuba.java
class Main {
  public static void main(String[] args) {
    System.out.print(karatsuba (923,1723));
  }
  private static int karatsuba(int x, int y){
    if (String.valueOf(x).length()==1 || (String.valueOf(y).length()==1)){
		return x*y;
    }
	  else{
		int n = Math.max(String.valueOf(x).length(),String.valueOf(y).length());
		int z = n / 2;
		int a = x /(int) Math.pow(10, z);
		int b = x %(int) Math.pow(10, z);
		int c = y /(int) Math.pow(10, z);
		int d = y %(int) Math.pow(10, z);
		int ac = karatsuba(a,c);
		int bd = karatsuba(b,d);
		int adbc = karatsuba(a+b,c+d) - ac - bd;
		return ac * (int)Math.pow(10, (2*z)) + (adbc * (int)Math.pow(10, z)) + bd;
    }
  }
}

File: Algorithms/Mathematics/Kotlin/karatsuba.kt
fun main(args:Array<String>) {
    print(karatsuba(923, 1723))
  }
fun karatsuba(x:Int, y:Int):Int {
    if ((x).toString().length == 1 || ((y).toString().length == 1))
    {
      return x * y
    }
    else
    {
      val n = Math.max((x).toString().length, (y).toString().length)
      val z = n / 2
      val a = x / Math.pow(10.0, z.toDouble()).toInt()
      val b = x % Math.pow(10.0, z.toDouble()).toInt()
      val c = y / Math.pow(10.0, z.toDouble()).toInt()
      val d = y % Math.pow(10.0, z.toDouble()).toInt()
      val ac = karatsuba(a, c)
      val bd = karatsuba(b, d)
      val adbc = karatsuba(a + b, c + d) - ac - bd
      return ac * Math.pow(10.0, (2 * z).toDouble()).toInt() + (adbc * Math.pow(10.0, z.toDouble()).toInt()) + bd
    }
 }